optimum/habana/accelerate/utils/transformer_engine.py

```diff
@@ -0,0 +1,137 @@
+
+
+def is_fp8_available():
+    return has_transformer_engine
+
+
+def _convert_model(model, to_transformer_engine=True, _convert_linear=True):
+    """
+    Recursively converts the linear and layernorm layers of a model to their `transformers_engine` counterpart.
+    """
+    if not is_fp8_available():
+        raise ImportError("Using `convert_model` requires transformer_engine to be installed.")
+    for name, module in model.named_children():
+        if type(module) == PEFTLinear and to_transformer_engine and _convert_linear:
+            LoRALinear.replace_forward(module)
+        if isinstance(module, torch.nn.Linear) and not type(module) == PEFTLinear and to_transformer_engine and _convert_linear:
+            has_bias = module.bias is not None
+            te_module = te.Linear(
+                module.in_features, module.out_features, bias=has_bias, params_dtype=module.weight.dtype, skip_weight_param_allocation=True
+            )
+            te_module.weight = module.weight
+
+            if has_bias:
+                te_module.bias = module.bias
+
+            setattr(model, name, te_module)
+        elif isinstance(module, te.Linear) and not to_transformer_engine and _convert_linear:
+            has_bias = module.bias is not None
+            new_module = torch.nn.Linear(
+                module.in_features, module.out_features, bias=has_bias, params_dtype=module.weight.dtype
+            )
+            new_module.weight.copy_(module.weight)
+            if has_bias:
+                new_module.bias.copy_(module.bias)
+
+            setattr(model, name, new_module)
+        else:
+            _convert_model(module, to_transformer_engine=to_transformer_engine, _convert_linear=_convert_linear)
+
+
+def has_transformer_engine_layers(model):
+    """
+    Returns whether a given model has some `transformer_engine` layer or not.
+    """
+    if not is_fp8_available():
+        raise ImportError("Using `has_transformer_engine_layers` requires transformer_engine to be installed.")
+    for m in model.modules():
+        if isinstance(m, (te.Linear)):
+            return True
+    return False
+
+def convert_model(model):
+    if not has_transformer_engine_layers(model):
+        with torch.no_grad():
+            _convert_model(model)
+        model._converted_to_transformer_engine = True
+    return model
+
+def get_fp8_format(fp8_format):
+    if fp8_format == "E5M2":
+        return te.recipe.Format.E5M2
+    elif fp8_format == "HYBRID":
+        return te.recipe.Format.HYBRID
+    else:
+        raise ValueError
+
+def get_fp8_recipe(fp8_config):
+    fp8_config = dict(fp8_config) if fp8_config is not None else {}
+    if "fp8_format" in fp8_config:
+        fp8_config['fp8_format'] = get_fp8_format(fp8_config['fp8_format'])
+    fp8_recipe_handler = te.recipe.DelayedScaling(**fp8_config)
+    fp8_recipe_handler.backend = "TE"
+    return fp8_recipe_handler
+
+class FP8ContextWrapper:
+    def __init__(self, ctx, fp8_recipe):
```

Comment:
Add comments explaining what class is doing.

Reply:
Done